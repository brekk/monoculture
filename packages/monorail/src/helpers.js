import {
  __ as $,
  any,
  ap,
  curry,
  defaultTo,
  filter,
  find,
  findLast,
  head,
  last,
  length,
  lt,
  map,
  pipe,
  propOr,
  reduce,
  slice,
  test,
} from 'ramda'
import { trace } from 'xtrace'

/**
 * @pageSummary Built-in helpers for making custom plugins more robust
 * @page helpers
 */

export const getBody = propOr([], 'body')

export const bodyTest = curry((fn, file, needle) =>
  pipe(getBody, fn(pipe(last, test(needle))))(file)
)
export const _reduce = curry((file, fn, initial) =>
  pipe(getBody, reduce(fn, initial))(file)
)

/**
 * Use this helper to test a regex that matches against any single incidence on any line
 * @name any
 * @example
 * ```js
 * const plugin = {
 *   name: 'get-const',
 *   fn: (state, file, { any }) => any(/const/)
 * }
 * export default plugin
 * ```
 */
export const _any = bodyTest(any)

/**
 * Use this helper to test a regex that matches and filters against every line
 * @name lines
 * @example
 * ```js
 * export default {
 *   name: 'unexported-consts',
 *   fn: (state, file, { lines }) => lines(/^const/)
 * }
 * ```
 */
export const onLines = curry((file, needle) =>
  pipe(bodyTest(filter, file), map(head))(needle)
)

/**
 * Use this helper to test a regex that matches and finds the first matching line
 * @name line
 * @example
 * ```js
 * export default {
 *   name: 'exported-default',
 *   fn: (state, file, { line }) => line(/export default/g)
 * }
 * ```
 */
export const onLine = curry((file, needle) =>
  pipe(bodyTest(find, file), defaultTo([-1]), head)(needle)
)

/**
 * Use this helper to test a regex that matches and finds the last matching line
 * @name lastLine
 * @example
 * ```js
 * export default {
 *   name: 'exported-last',
 *   fn: (state, file, { lastLine }) => lastLine(/export/g)
 * }
 * ```
 */
export const onLastLine = curry((file, needle) =>
  pipe(bodyTest(findLast, file), defaultTo([-1]), head)(needle)
)

/**
 * Use this helper to select content between two repeating regular expressions
 * @name between
 * @example
 * ```js
 * export default {
 *   name: 'expanded-imports',
 *   fn: (state, file, { between }) => between(/^import/, /from (.*)/)
 * }
 * ```
 */
export const selectBetween = curry((file, start, end) =>
  pipe(
    z => [z],
    ap([onLine($, start), onLastLine($, end), getBody]),

    ([a, z, body]) =>
      a === -1 || z === -1
        ? []
        : filter(([line]) => a <= line && line <= z)(body)
  )(file)
)

/**
 * Use this helper to select all content between two repeating regular expressions
 * @name selectAll
 * @example
 * ```js
 * export default {
 *   name: 'expanded-imports',
 *   fn: (state, file, { selectAll }) => selectAll(/^import/, /from (.*)/)
 * }
 * ```
 */
export const selectAll = curry((file, start, end) =>
  pipe(
    getBody,
    reduce(
      ({ active, all, current }, [line, content]) => {
        const testContent = test($, content)
        const checkStart = testContent(start)
        const stillActive = checkStart || active
        if (stillActive) {
          const newCurrent = current.concat([line, content])
          const checkEnd = testContent(end)
          if (checkEnd) {
            return {
              active: false,
              all: all.concat([newCurrent]),
              current: [],
            }
          }
          return {
            active: !checkEnd,
            all,
            current: newCurrent,
          }
        }
        return {
          active: false,
          all: all.concat([current]),
          current: [],
        }
      },
      { active: false, all: [], current: [] }
    ),
    propOr([], 'all'),
    filter(pipe(length, lt(0)))
  )(file)
)

export const _filter = bodyTest(filter)

export const makeFileHelpers = file => ({
  any: _any(file),
  onLines: onLines(file),
  onLine: onLine(file),
  onLastLine: onLastLine(file),
  filter: _filter(file),
  between: selectBetween(file),
  selectAll: selectAll(file),
  reduce: _reduce(file),
})
export const _getConfigFrom = curry((name, c) => c?.config?.[name])

export const makePluginHelpers = curry((state, plugin) => ({
  config: _getConfigFrom(plugin.name, state),
}))
