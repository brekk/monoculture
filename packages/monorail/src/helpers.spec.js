import { makeHelpers } from './helpers'

const FILE = {
  body: `import {
  __ as $,
  any,
  ap,
  curry,
  defaultTo,
  filter,
  find,
  findLast,
  head,
  last,
  length,
  lt,
  map,
  pipe,
  propOr,
  reduce,
  slice,
  test,
} from 'ramda'
import { trace } from 'xtrace'

export const getBody = propOr([], 'body')

export const bodyTest = curry((fn, file, needle) =>
  pipe(getBody, fn(pipe(last, test(needle))))(file)
)
export const _reduce = curry((file, fn, initial) =>
  pipe(getBody, reduce(fn, initial))(file)
)
`
    .split('\n')
    .map((v, k) => [k + 1, v]),
}
describe('makeHelpers', () => {
  const helpers = makeHelpers(FILE)
  it('any', () => {
    const some = helpers.any(/^import/)
    expect(some).toMatchSnapshot()
  })
  it('lines', () => {
    const lines = helpers.onLines(/^import/)
    expect(lines).toMatchSnapshot()
  })
  it('line', () => {
    const line = helpers.onLine(/ramda/)
    expect(line).toMatchSnapshot()
  })
  it('filter', () => {
    const filter = helpers.filter(/export/g)
    expect(filter).toMatchSnapshot()
  })
  it('between', () => {
    const between = helpers.between(/^import/, /from (.*)$/)
    expect(between).toMatchSnapshot()
    const twixtNothing = helpers.between(/aaa/, /bbb/)
    expect(twixtNothing).toMatchSnapshot()
  })
  it('selectAll', () => {
    const selectAll = helpers.selectAll(/^import/, /from (.*)$/)
    expect(selectAll).toMatchSnapshot()
  })
  it('reduce', () => {
    const reduce = helpers.reduce((agg, [, content]) => agg.concat(content), [])
    expect(reduce).toMatchSnapshot()
  })
})
