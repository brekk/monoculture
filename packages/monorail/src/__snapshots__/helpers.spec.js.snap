// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeHelpers any 1`] = `true`;

exports[`makeHelpers between 1`] = `
[
  [
    1,
    "import {",
  ],
  [
    2,
    "  __ as $,",
  ],
  [
    3,
    "  any,",
  ],
  [
    4,
    "  ap,",
  ],
  [
    5,
    "  curry,",
  ],
  [
    6,
    "  defaultTo,",
  ],
  [
    7,
    "  filter,",
  ],
  [
    8,
    "  find,",
  ],
  [
    9,
    "  findLast,",
  ],
  [
    10,
    "  head,",
  ],
  [
    11,
    "  last,",
  ],
  [
    12,
    "  length,",
  ],
  [
    13,
    "  lt,",
  ],
  [
    14,
    "  map,",
  ],
  [
    15,
    "  pipe,",
  ],
  [
    16,
    "  propOr,",
  ],
  [
    17,
    "  reduce,",
  ],
  [
    18,
    "  slice,",
  ],
  [
    19,
    "  test,",
  ],
  [
    20,
    "} from 'ramda'",
  ],
  [
    21,
    "import { trace } from 'xtrace'",
  ],
]
`;

exports[`makeHelpers between 2`] = `[]`;

exports[`makeHelpers filter 1`] = `
[
  [
    23,
    "export const getBody = propOr([], 'body')",
  ],
  [
    25,
    "export const bodyTest = curry((fn, file, needle) =>",
  ],
  [
    28,
    "export const _reduce = curry((file, fn, initial) =>",
  ],
]
`;

exports[`makeHelpers line 1`] = `20`;

exports[`makeHelpers lines 1`] = `
[
  1,
  21,
]
`;

exports[`makeHelpers reduce 1`] = `
[
  "import {",
  "  __ as $,",
  "  any,",
  "  ap,",
  "  curry,",
  "  defaultTo,",
  "  filter,",
  "  find,",
  "  findLast,",
  "  head,",
  "  last,",
  "  length,",
  "  lt,",
  "  map,",
  "  pipe,",
  "  propOr,",
  "  reduce,",
  "  slice,",
  "  test,",
  "} from 'ramda'",
  "import { trace } from 'xtrace'",
  "",
  "export const getBody = propOr([], 'body')",
  "",
  "export const bodyTest = curry((fn, file, needle) =>",
  "  pipe(getBody, fn(pipe(last, test(needle))))(file)",
  ")",
  "export const _reduce = curry((file, fn, initial) =>",
  "  pipe(getBody, reduce(fn, initial))(file)",
  ")",
  "",
]
`;

exports[`makeHelpers selectAll 1`] = `
[
  [
    1,
    "import {",
    2,
    "  __ as $,",
    3,
    "  any,",
    4,
    "  ap,",
    5,
    "  curry,",
    6,
    "  defaultTo,",
    7,
    "  filter,",
    8,
    "  find,",
    9,
    "  findLast,",
    10,
    "  head,",
    11,
    "  last,",
    12,
    "  length,",
    13,
    "  lt,",
    14,
    "  map,",
    15,
    "  pipe,",
    16,
    "  propOr,",
    17,
    "  reduce,",
    18,
    "  slice,",
    19,
    "  test,",
    20,
    "} from 'ramda'",
  ],
  [
    21,
    "import { trace } from 'xtrace'",
  ],
]
`;
